File: tests/workflow/producer/rust/src/producer/events/events_eventb.rs
use super::{identification, producer::Control, protocol, scope::AnonymousScope, Context};
use crate::stat::Alias;
use crate::test::samples;
use clibri::server;
use std::str::FromStr;
use uuid::Uuid;

type BroadcastGroupAStructA = (Vec<Uuid>, protocol::GroupA::StructA);
type BroadcastGroupAStructB = (Vec<Uuid>, protocol::GroupA::StructB);
type BroadcastGroupBStructA = (Vec<Uuid>, protocol::GroupB::StructA);

#[allow(unused_variables)]
pub async fn emit<E: server::Error, C: server::Control<E>>(
    event: protocol::Events::EventB,
    scope: &mut AnonymousScope<'_, E, C>,
) -> Result<
    (
        BroadcastGroupAStructA,
        BroadcastGroupAStructB,
        BroadcastGroupBStructA,
    ),
    String,
> {
    let uuid = match Uuid::from_str(&event.uuid) {
        Ok(uuid) => uuid,
        Err(err) => {
            return Err(format!("Fail to parse uuid {}: {:?}", event.uuid, err));
        }
    };
    scope.context.inc_stat(uuid, Alias::GroupAStructA);
    scope.context.inc_stat(uuid, Alias::GroupAStructB);
    scope.context.inc_stat(uuid, Alias::GroupBStructA);
    Ok((
        (vec![uuid], samples::group_a::struct_a::get()),
        (vec![uuid], samples::group_a::struct_b::get()),
        (vec![uuid], samples::group_b::struct_a::get()),
    ))
}

File: tests/workflow/producer/typescript/src/producer/events/events.eventb.ts
import { Filter, Producer, Context, Protocol } from "../implementation/events";
import { Output } from "../implementation/events/events.eventb";
import { Alias } from "../../stat";

export function emit(
	event: Protocol.Events.EventB,
	filter: Filter,
	context: Context,
	producer: Producer
): Promise<Output> {
	const stat = context.getStat(event.uuid);
	stat.case(Alias.GroupAStructA);
	stat.case(Alias.GroupAStructB);
	stat.case(Alias.GroupBStructA);
	return Promise.resolve(
		new Output()
			.broadcast([event.uuid])
			.GroupAStructA(Protocol.GroupA.StructA.defaults())
			.broadcast([event.uuid])
			.GroupAStructB(Protocol.GroupA.StructB.defaults())
			.broadcast([event.uuid])
			.GroupBStructA(Protocol.GroupB.StructA.defaults())
	);
}
